sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -190,82 +190,45 @@ public AccessTokenCache(TokenCredential credential, TimeSpan tokenRefreshOffset,
                 _tokenRefreshRetryDelay = tokenRefreshRetryDelay;
             }
 
-            public async ValueTask<string> GetHeaderValueAsync(HttpMessage message, TokenRequestContext context, bool async)
+            public async ValueTask<string> GetAuthHeaderValueAsync(HttpMessage message, TokenRequestContext context, bool async)
             {
-                bool getTokenFromCredential;
-                TaskCompletionSource<HeaderValueInfo> headerValueTcs;
-                TaskCompletionSource<HeaderValueInfo>? backgroundUpdateTcs;
+                bool shouldRefreshFromCredential;
                 int maxCancellationRetries = 3;
 
                 while (true)
                 {
-                    (headerValueTcs, backgroundUpdateTcs, getTokenFromCredential) = GetTaskCompletionSources(context);
-                    HeaderValueInfo info;
-                    if (getTokenFromCredential)
+                    shouldRefreshFromCredential = RefreshTokenRequestState(context, out TokenRequestState localState);
+                    AuthHeaderValueInfo headerValueInfo;
+
+                    if (shouldRefreshFromCredential)
                     {
-                        if (backgroundUpdateTcs != null)
+                        if (localState.BackgroundTokenUpdateTcs != null)
                         {
-                            if (async)
-                            {
-                                info = await headerValueTcs.Task.ConfigureAwait(false);
-                            }
-                            else
-                            {
-#pragma warning disable AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
-                                info = headerValueTcs.Task.EnsureCompleted();
-#pragma warning restore AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
-                            }
-                            _ = Task.Run(() => GetHeaderValueFromCredentialInBackgroundAsync(backgroundUpdateTcs, info, context, async));
-                            return info.HeaderValue;
+                            headerValueInfo = await localState.GetCurrentHeaderValue(async).ConfigureAwait(false);
```

Comment:
Curious: does removing the `async` switch here have us doing sync-over-async in a sync context?

Reply:
Got it.  Dunno how I missed that!